File: Memory-JAVA/app/src/main/java/com/andraganoid/memory_java/MainActivity.java
package com.andraganoid.memory_java;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;


import com.andraganoid.memory_java.databinding.ActivityMainBinding;

import java.util.List;


public class MainActivity extends AppCompatActivity {

    MemoryViewModel memoryViewModel;
    GridView memoGridView;
    GridMemoryAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Display display = getWindowManager().getDefaultDisplay();
        DisplayMetrics dm = new DisplayMetrics();
        display.getMetrics(dm);

        memoryViewModel = ViewModelProviders.of(this).get(MemoryViewModel.class);
        ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);//.vmodel = myWeatherViewModel

        binding.setResult(memoryViewModel.memoryResult);
        binding.setViewModel(memoryViewModel);

        memoGridView = findViewById(R.id.fields_grid_view);
        mAdapter = new GridMemoryAdapter(this, (int) (dm.heightPixels * .12));
        memoGridView.setAdapter(mAdapter);

        memoryViewModel.getMemoryList().observe(this, new Observer <List <Field>>() {
            @Override
            public void onChanged(List <Field> fields) {
                mAdapter.setFields(fields);
            }
        });


        memoGridView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView <?> parent, View view, int position, long id) {
                memoryViewModel.fieldClicked(position);
            }
        });
    }

    @Override
    protected void onPause() {
        super.onPause();
        memoryViewModel.stopStopwatch();
    }

}

File: Memory-KOTLIN/app/src/main/java/com/andraganoid/memory_kotlin/MainActivity.kt
package com.andraganoid.memory_kotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    lateinit var memoryViewModel: MemoryViewModel
    val ALL_SOLVED = 8

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        newGameButton.visibility = View.VISIBLE;
        var display = getWindowManager().getDefaultDisplay()
        var dm = DisplayMetrics();
        display.getMetrics(dm);

        val mAdapter = GridMemoryAdapter(this, (dm.heightPixels * .12).toInt())
        fieldsGridView.adapter = mAdapter

        fieldsGridView.setOnItemClickListener { parent, view, position, id ->
            memoryViewModel.fieldClicked(position);
        }

        memoryViewModel = ViewModelProviders.of(this).get(MemoryViewModel::class.java)
        memoryViewModel.memoryList.observe(this,
            Observer<List<Field>> { fields ->
                mAdapter.setFields(fields)
            })

        memoryViewModel.memoryResults.observe(this,
            Observer<Any> { result -> setResults(result as Result) })

        newGameButton.setOnClickListener {
            it.visibility = View.INVISIBLE
            memoryViewModel.newGame();
        }

    }

    override fun onPause() {
        super.onPause()
        memoryViewModel.stopTimer();
    }

    fun setResults(result: Result) {
        bestMovesValue.setText(result.bestMoves.toString())
        bestTimeValue.setText(result.bestTimeInSeconds())
        currentMovesValue.setText(result.currentMoves.toString())
        currentTimeValue.setText(result.currentTimeInSeconds())
        if (result.solved == ALL_SOLVED) {
            newGameButton.visibility = View.VISIBLE
        }
    }

}


